PDEP-1: Purpose and guidelines for pandas enhancement proposals (#47444)
pandas-dev/pandas: web/pandas/pdeps/accepted/0001-purpose-and-guidelines.md

# PDEP-1: Purpose and guidelines

> - Date 21 June 2022

Reviewer: ```suggestion
- Date: 21 June 2022
```

discussion and require detailed documentation before being implemented.

PDEP are appropriate for user facing changes, internal changes and organizational
discussions. Examples of topics worth a PDEP could include moving a module from
pandas to a separate repository, a refactoring of the pandas block manager or
a proposal of a new code of conduct.

## PDEP guidelines

### Target audience

A PDEP is a public document available to anyone, but the main stakeholders to
consider when writing a PDEP are:

- The core development team, who will have the final decision on whether a PDEP
  is approved or not
- Developers of pandas and other related projects, and experienced users. Their
  feedback is highly encouraged and appreciated, to make sure all points of views
  are taken into consideration
- The wider pandas community, in particular users, who may or may not have feedback
  on the proposal, but should know and be able to understand the future direction of
  the project

### PDEP authors

Anyone can propose a PDEP, but in most cases developers of pandas itself and related
projects are expected to author PDEPs. If you are unsure if you should be opening

> an issue or creating a PDEP, it's probably save to start by

Reviewer: ```suggestion
an issue or creating a PDEP, it's probably safe to start by
```

[opening an issue](https://github.com/pandas-dev/pandas/issues/new/choose), which can
be eventually moved to a PDEP.

### Workflow

#### Submitting a PDEP

Proposing a PDEP is done by creating a PR adding a new file to `web/pdeps/accepted/`.
The file is a markdown file, you can use `web/pdeps/accepted/0001.md` as a reference
for the expected format.

By default, we expect a PDEP will be accepted, so the PR of a PDEP should be done
in the `accepted` directory and contain `Status: Accepted`. If a PDEP is finally
rejected, its status and directory will be updated by the core team before merging,
once the decision is made. Please make sure you select the option
`Allow edits and access to secrets by maintainers` when opening the PR.

#### Accepted PDEP

A PDEP will be accepted by the core development team, and decisions will be made
based on the [pandas governance document](https://github.com/pandas-dev/pandas-governance/blob/master/governance.md).

> Once a PDEP is accepted, it is unclear how the implementation will happen. The

Reviewer 2: ```suggestion
Once a PDEP is accepted, any contributions can be made toward the implementing the PDEP with an open-ended completion timeline . The
```

IMO the former wording doesn't sound that encouraging to propose PDEPs that might not be implemented

pandas project development, with a mix of volunteers and developers paid from
different sources, and development priorities are difficult to understand or
forecast. For companies, institutions or individuals with interest in seeing a
PDEP being implemented, or to in general see progress to the pandas roadmap,
please check how you can help in the [contributing page](/contribute.html).

#### Implemented PDEP

> Once a PDEP is implemented and available in the main branch of pandas, its

Reviewer 2: Should the release version be included too to note which version of pandas has the complete implementation?

Author: Great idea, I included it.

a proposal of a new code of conduct.

## PDEP guidelines

### Target audience

A PDEP is a public document available to anyone, but the main stakeholders to
consider when writing a PDEP are:

- The core development team, who will have the final decision on whether a PDEP
  is approved or not
- Developers of pandas and other related projects, and experienced users. Their
  feedback is highly encouraged and appreciated, to make sure all points of views
  are taken into consideration
- The wider pandas community, in particular users, who may or may not have feedback
  on the proposal, but should know and be able to understand the future direction of
  the project

### PDEP authors

Anyone can propose a PDEP, but in most cases developers of pandas itself and related
projects are expected to author PDEPs. If you are unsure if you should be opening
an issue or creating a PDEP, it's probably safe to start by
[opening an issue](https://github.com/pandas-dev/pandas/issues/new/choose), which can
be eventually moved to a PDEP.

### Workflow

#### Submitting a PDEP

Proposing a PDEP is done by creating a PR adding a new file to `web/pdeps/accepted/`.

> The file is a markdown file, you can use `web/pdeps/accepted/0001.md` as a reference

Author: I'd personally leave that for later, once we need it. Those sound like good ideas to me, but I wouldn't implement them initially, I would wait to see how things work first. If we merge like PDEP per month, I think checking the last PDEP number before merging is easier than a system to autogenerate them. And for a template, I'd wait to have few actual PDEPs before deciding if it helps, or if PDEPs are too different from one to another.

Does it make sense to you to start simple and iterate later as we have more experience?

for the expected format.

> By default, we expect a PDEP will be accepted, so the PR of a PDEP should be done

Reviewer 3: Do we want to also have a public comment period on PDEP's like tensorflow has with their RFC's?

We should probably also clarify when voting happens(define what proportion of core team members need to consider a PDEP ready before voting like @mroeschke stated) and how long core team members have to vote.

Reviewer: > But again, I'd start by seeing how things work, and if we see PDEPs keep open for too long, and seems like having a timeframe should help, surely worth giving it a try.

I agree, just having a PDEP is akin to a new process, so having a process for the PDEP is maybe like introducing another meta process at the same time. So in response to this comment and others related to the process, I agree with @datapythonista that to begin with we just discuss whether we want to use PDEPs, what we want out of a PDEP,  and roughly what it should contain and iterate on the gaps over time. The only caveat here, is that once someone has spent effort preparing a PDEP and it is approved that we don't have "blockers" implementing, so that does affect the approval process to some degree.

Reviewer 4: i agree - we need a voting process here - we could emulate that of NEP which is pretty simple

Reviewer 5: > we need a voting process here

I think what Marc proposed above is to defer a discussion about the decision process (basically our governance model) for a subsequent discussion.

(which sounds good to me to keep the discussion manageable)

> in the `accepted` directory and contain `Status: Accepted`. If a PDEP is finally

Author: Sounds good to me.

Regarding merging PDEPs as draft, feels like it could make discussions more difficult. Like where to have the discussion once the PR is merged. Via new PRs? Or in a second PR, not being able to comment in parts not in the diff. But I guess in some cases it can be useful, maybe a PR with different parts, and merging as a Draft once the first part has been discussed, and before working on other parts.

I'll update the document if there are no objections to using Under discussion status (or Draft if people prefer this name) until PR is Accepted. And to consider merging PRs before they are accepted when it's useful.

> in the `accepted` directory and contain `Status: Accepted`. If a PDEP is finally

Reviewer 5: > Regarding merging PDEPs as draft, feels like it could make discussions more difficult.

We will probably have to see how it goes in practice. For many PDEPs a single PR with proposal, discussion, acceptance all in one will quite likely be sufficient. 
But I can imagine for larger/longer-running discussions (I am thinking eg about the copy/view proposal), that it could be useful to already have a hosted version of PDEP that is still under discussion. 

As some references: numpy seems to already merge NEP PRs if they are still under discussion (eg https://github.com/numpy/numpy/pull/18456), but they also have substantial discussion on the mailing list. Similarly in Python, PEPs are merged quickly, but also there the discussion is mostly happening on mailing list or discourse. That's of course different with the current way in pandas where most discussion typically happens on GitHub.

available will also be included in the PDEP.

#### Rejected PDEP

A PDEP can be rejected when the final decision is that its implementation is
not the best for the interests of the project. They are as useful as accepted
PDEPs, since there are discussions that are worth having, and decisions about
changes to pandas being made. They will be merged with `Status: Rejected`, so
there is visibility on what was discussed and what was the outcome of the
discussion. A PDEP can be rejected for different reasons, for example good ideas
that aren't backward-compatible, and the breaking changes aren't considered worth
implementing.

#### Invalid PDEP

For submitted PDEPs that do not contain proper documentation, are out of scope, or
are not useful to the community for any other reason, the PR will be closed after
discussion with the author, instead of merging them as rejected. This is to not
add noise to the list of rejected PDEPs, which should contain documentation as
good as an accepted PDEP, but where the final decision was to not implement the changes.

## Evolution of this PDEP

While most PDEPs aren't expected to change after accepted, in some cases like this

> PDEP, they will be updated to contain its latest version, if things evolve. A log

Reviewer 3: We should probably clarify when a PDEP can be revised v.s. when a new PDEP has to be submitted superseding the old PDEP.

A PDEP is a public document available to anyone, but the main stakeholders to
consider when writing a PDEP are:

- The core development team, who will have the final decision on whether a PDEP
  is approved or not
- Developers of pandas and other related projects, and experienced users. Their
  feedback is highly encouraged and appreciated, to make sure all points of views
  are taken into consideration
- The wider pandas community, in particular users, who may or may not have feedback
  on the proposal, but should know and be able to understand the future direction of
  the project

### PDEP authors

Anyone can propose a PDEP, but in most cases developers of pandas itself and related
projects are expected to author PDEPs. If you are unsure if you should be opening
an issue or creating a PDEP, it's probably safe to start by
[opening an issue](https://github.com/pandas-dev/pandas/issues/new/choose), which can
be eventually moved to a PDEP.

### Workflow

#### Submitting a PDEP

Proposing a PDEP is done by creating a PR adding a new file to `web/pdeps/accepted/`.
The file is a markdown file, you can use `web/pdeps/accepted/0001.md` as a reference
for the expected format.

> By default, we expect a PDEP will be accepted, so the PR of a PDEP should be done

Reviewer 4: i am not sure this is true - sure a lot will easily be accepted but some might be controversial and ultimately not accepted

Reviewer: yeah. maybe just qualify this by changing `By default, we expect a PDEP will be accepted` to something like `By default, we expect a PDEP (with some preliminary discussion) will be accepted` 

before this text we have...

> A PDEP should be used for changes that are not
immediate and not obvious, and are expected to require a significant amount of
discussion and require detailed documentation before being implemented.

so the PEP is part of the detailed documentation and an issue should perhaps be the initial part of the discussion. (just as I would normally expect say a bug fix PR to have an associated issue)

and we also have 

> If you are unsure if you should be opening
an issue or creating a PDEP, it's probably safe to start by
[opening an issue](https://github.com/pandas-dev/pandas/issues/new/choose), which can
be eventually moved to a PDEP.

I think that maybe issues should **always** be opened before submitting a PEP, either as a specific issue or an existing issue concluding that a PEP is required. Maybe we also need to ensure that `@pandas-dev/pandas-core` is also notified in these cases.

If we have some form of gate for opening a PEP, then most should be accepted.

In general, more than one approval will be needed before the PR is merged. And
there should not be any `Request changes` review at the time of merging.

Once a PDEP is accepted, any contributions can be made toward the implementing the PDEP with an open-ended completion timeline . The
pandas project development, with a mix of volunteers and developers paid from
different sources, and development priorities are difficult to understand or
forecast. For companies, institutions or individuals with interest in seeing a
PDEP being implemented, or to in general see progress to the pandas roadmap,
please check how you can help in the [contributing page](/contribute.html).

#### Implemented PDEP

Once a PDEP is implemented and available in the main branch of pandas, its
status will be changed to implemented, so there is visibility that the PDEP
is not part of the roadmap and future plans, but a change that it already
happened. The first pandas version in which the PDEP implementation is
available will also be included in the PDEP.

#### Rejected PDEP

A PDEP can be rejected when the final decision is that its implementation is

> not the best for the interests of the project. They are as useful as accepted

Reviewer 6: This is to contrast rejected PDEPs against accepted PDEPs; i.e. "even rejected PDEPs are useful". I think the wording is okay, but maybe "Rejected PDEPs are just as useful as accepted PDEPs..." would make it more clear?

> not the best for the interests of the project. They are as useful as accepted

Reviewer 5: Ah, I missed the first "as" in the sentence, so I read it as "they are useful as accepted PDEP", instead of "they are _as_ useful as accepted PDEPs", hence the confusion.
